src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs

```diff
@@ -251,15 +251,18 @@ private ImmutableArray<TypeParameterSymbol> MakeTypeParameters(DiagnosticBag dia
         /// <summary>
         /// Returns the constraint clause for the given type parameter.
         /// </summary>
-        internal TypeParameterConstraintClause GetTypeParameterConstraintClause(int ordinal)
+        internal TypeParameterConstraintClause GetTypeParameterConstraintClause(bool canIgnoreNullableContext, int ordinal)
         {
             var clauses = _lazyTypeParameterConstraints;
-            if (clauses.IsDefault)
+            if (!clauses.IsSet(canIgnoreNullableContext))
             {
                 var diagnostics = DiagnosticBag.GetInstance();
-                if (ImmutableInterlocked.InterlockedInitialize(ref _lazyTypeParameterConstraints, MakeTypeParameterConstraints(diagnostics)))
+                if (ImmutableInterlocked.InterlockedCompareExchange(ref _lazyTypeParameterConstraints, MakeTypeParameterConstraints(canIgnoreNullableContext, diagnostics), clauses) == clauses)
```

Comment:
>InterlockedCompareExchange [](start = 41, length = 26)

Is there a possibility of a race here? If a thread with ```canIgnoreNullableContext==true``` completes first, is the thread with ```canIgnoreNullableContext==true``` going to return expected value?

Reply:
Good catch, thanks.

---
In reply to: [493779200](https://github.com/dotnet/roslyn/pull/47968#discussion_r493779200) [](ancestors = 493779200)